argsman, cli: GNU-style command-line option parsing (allows options after non-option arguments)

pablomartin4btc opened the pull request:
This PR simplifies the argument parsing logic and resolves practical issues for `bitcoin-cli` and similar binaries (e.g. `bitcoin-wallet`) making their usage more consistent and predictable, enabling valid patterns such as mixing commands with wallet- or RPC-specific options—without relying on argument ordering quirks.

<details>
<summary>What's changed in ArgsManager that allows the above.</summary>

- Previously, `ArgsManager` stopped interpreting options once a non-option argument was encountered (non-GNU parsing). For example:

        bitcoind -a -b param -c

  would ignore `-c` because it followed the non-option argument `param`. In some cases '-c' is interpreted as an argument for 'param', producing unexpected behavior (eg empty arrays for listtransactions command).

- This PR changes the behavior so that options following non-option argument are recognized and processed (GNU parsing). This is needed for workflows for tools like CLI where command arguments may be followed by wallet- or RPC-specific
options, e.g.:

        bitcoin-cli -rpcwallet="" listtransactions -rpcwallet=mywallet
    
        bitcoin-cli getaddressinfo <address> -rpcwallet=mywallet

- (other binaries could also benefit from this change,  e.g. `bitcoin-wallet`)

        bitcoin-wallet create -wallet=newname

- Invalid options passed after non-options will be treated as RPC arguments that begin with a `-` character (so no
behavioural change against `master`), e.g.:

        bitcoin-cli -regtest -datadir=/tmp/btc createwallet -mywallet
        {
            "name": "-mywallet"
        }

</details>

<details><summary>Before and after outputs running RPC commands from <code>CLI</code>/ <code>bitcoin-cli</code>.</summary><ul>
     <li><details><summary>wallet commands using <code>CLI</code>:</summary><ul> 
           <li><details><summary><code>getaddressinfo</code></summary><ul>
           <li>before/ <code>master</code>

```
./build_master/bin/bitcoin-cli -regtest -datadir=/tmp/btc getaddressinfo bcrt1qrnz5xlcwz5d85n3fag6v0vc4lryjc90jhqtpsh -rpcwallet=nonExistent
error message:
getaddressinfo "address"
...
```
</li>
           <li>after

```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc getaddressinfo bcrt1qrnz5xlcwz5d85n3fag6v0vc4lryjc90jhqtpsh -rpcwallet=nonExistent
error code: -18
error message:
Requested wallet does not exist or is not loaded
```

</li>
           </ul></details></li>
           <li><details><summary><code>listtransactions</code></summary><ul>
           <li>before/ <code>master</code>

```
./build_master/bin/bitcoin-cli -regtest -datadir=/tmp/btc listtransactions -rpcwallet=nonExistent
[
]
```

</li>
           <li>after


```
./build/bin/bitcoin-cli -regtest -datadir=/tmp/btc listtransactions -rpcwallet=nonExistent
error code: -18
error message:
Requested wallet does not exist or is not loaded
```

</li>
           </ul></details></li>
           <li><details><summary>any command that needs an option to be specifed, it can be added at the end directly instead of before the command</summary><ul>
           <li>before/ <code>master</code>

```
./build_master/bin/bitcoin-cli -regtest -datadir=/tmp/btc listtransactions
error code: -19
error message:
Multiple wallets are loaded. Please select which wallet to use by requesting the RPC through the /wallet/<walletname> URI path. Or for the CLI, specify the "-rpcwallet=<walletname>" option before the command (run "bitcoin-cli -h" for help or "bitcoin-cli listwallets" to see which wallets are currently loaded).
```

```
./build_master/bin/bitcoin-cli -regtest -datadir=/tmp/btc listtransactions -rpcwallet=""
error code: -19
error message:
Multiple wallets are loaded. Please select which wallet to use by requesting the RPC through the /wallet/<walletname> URI path. Or for the CLI, specify the "-rpcwallet=<walletname>" option before the command (run "bitcoin-cli -h" for help or "bitcoin-cli listwallets" to see which wallets are currently loaded).
```

</li>
           <li>after

```
/build/bin/bitcoin-cli -regtest -datadir=/tmp/btc listtransactions -rpcwallet=""
[
  {
    "address": "bcrt1q5pqkrlfrp9rvsg43sxxagrees7r7ul6uyskhs4",
    "parent_descs": [
      "wpkh([76a11ab0/84h/1h/0h]tpubDCDCuyC1Rtm1weZ5kfBYtdchcmqa2t8db7M68wsPHbWhUrYnzFTrHu7ufVVRydt3FdWQy8iNmKbZYbgpWWcEKPAj896x53UbS6i2xRD1qct/0/*)#kuff3m9t"
    ],
    "category": "immature",
    "amount": 50.00000000,
    "label": "",
...
```

</li>
           </ul></details></li>
     </ul></details></li>
     <li><details><summary>other binaries could benefit from it, e.g. <code>bitcoin-wallet:</code></summary><ul> 
           <li><details><summary><code>create</code> command</summary><ul>
           <li>before/ <code>master</code>

```
./build_master/bin/bitcoin-wallet -regtest -datadir=/tmp/btc create
Wallet name must be provided when creating a new wallet.
```
_(`-wallet` option has to be passed before the `create` command in order to work)_
```
./build_master/bin/bitcoin-wallet -regtest -datadir=/tmp/btc create -wallet=newWalletName
Error: Additional arguments provided (-wallet=newWalletName). Methods do not take arguments. Please refer to `-help`.
```

</li>
           <li>after

```
./build/bin/bitcoin-wallet -regtest -datadir=/tmp/btc create -wallet=newWallet
Topping up keypool...
Wallet info
===========
Name: newWallet
Format: sqlite
Descriptors: yes
Encrypted: no
HD (hd seed available): yes
Keypool Size: 8000
Transactions: 0
Address Book: 0
```
</li>
           </ul></details></li>
     </ul></details></li>
     </ul> <!-- End -->
</details>

-<ins>_**Notes**_</ins>:

- <ins>Backwards compatibility</ins>:
This PR is backward compatible in the sense that it allows new valid invocations that were previously rejected (failed or somehow ignored).
In `master`, any valid option following a command is treated as a positional argument. This can lead to unexpected RPC errors or help messages due to mismatched argument types. With this PR, those cases are parsed as proper options when applicable.

- `bitcoin-cli` continues to accept RPC arguments that begin with a `-` character as currently in `master` (check `createwallet` example in _"What's change in ArgsManager"_ section above).

- `bitcoin-cli` could distinguish between options that begin with single and double dashes, and treat options after the RPC method name that begin with double dashes as RPC named parameters (check example in 4th [commit](https://github.com/bitcoin/bitcoin/pull/33540/commits/a101bbb61fdf5563337334d63a10432432ee3625) message body).

- Refactored both `ArgsManager::ProcessOptionKey` (new function added in this PR) and `ArgsManager::ParseParameters` in separated commits making the code clearer and the PR much easier to follow and review.

- On Windows, command-specific options using the '/' prefix are now handled correctly when specified after a command.

- Updated `bitcoin-tx` to consume the command arguments already parsed by `ArgsManager`, avoiding re-processing raw `argv` in `CommandLineRawTx()`.

Additionally, this PR includes release notes documenting the GNU-style option parsing behaviour and the distinction between command-line options and RPC named parameters.

pablomartin4btc committed:
argsman, refactor: Factor out ProcessOptionKey from ParseParameters

- Centralizes normalization of option keys in `ProcessOptionKey`.
  This will help in next commit to reuse this function avoiding code
  duplication.
- Improves readability and makes behavior easier to reason about.

Functional and unit tests confirm no behavior change.

pablomartin4btc committed:
argsman, cli, test: Allow options after non-option arguments

Previously, ArgsManager stopped interpreting options once a
non-option argument was encountered (non-GNU parsing).

For example:

    bitcoind -a -b param -c

would ignore `-c` because it followed the non-option argument
`param`. In cases where some 'param' command receives parameters,
'-c' is interpreted as one of them producing unexpected behavior
(e.g. empty arrays for listtransactions command).

This commit changes the behavior so that options following
non-option argument are recognized and processed (GNU parsing).
This is needed for workflows for tools like CLI where command
arguments may be followed by wallet- or RPC-specific options, e.g.:

    bitcoin-cli listtransactions -rpcwallet=mywallet

    bitcoin-cli getaddressinfo <address> -rpcwallet=mywallet

    bitcoin-wallet create -wallet=newname

Also, invalid options passed after non-options will be treated
as RPC arguments that begin with a '-' (dash) character (so no
behavioural change against master), e.g.:

    bitcoin-cli -regtest -datadir=/tmp/btc createwallet -mywallet
    {
        "name": "-mywallet"
    }

pablomartin4btc committed:
cli, refactor, test: Use parsed command args from ArgsManager

Refactor bitcoin-cli to use command arguments provided by ArgsManager
instead of manually slicing argv.

- Introduce GetCommandArgs() helper.
- Remove local argv/argc skipping logic.
- Simplify CommandLineRPC argument collection.

This improves consistency in bitcoin-cli, avoiding duplicating parsing
logic.

Added a new functional test passing an option after a non-option
argument (GNU-style).

No behavior change intended.

w0xlt reviewed (commented):
Concept ACK

ryanofsky commented:
Haven't looked at code yet, so forgive if these are dumb questions, but could be nice if PR description addressed these:

- Is there a way to pass `bitcoin-cli` RPC arguments that begin with a `-` character? Can you escape these arguments with `--` for example?

- What's the PR's goal for backwards compatibility? Is the PR only allowing command lines that were disallowed before, and therefore fully backwards compatible? Or are there some command lines which used to be accepted which will now be interpreted differently?

- What are "numeric command arguments" referenced in the PR description? Maybe give an example. I wasn't aware that numbers were treated differently than other characters for separating options arguments, and I would expect interpretation of numeric values to happen at a later phase than the phase distinguishing options from non-option arguments.

ryanofsky commented:
Concept ACK on the idea. I think it probably makes sense to allow options arguments to follow non-option arguments even this change is not strictly backwards compatible, as long as there is some way to escape non-option arguments beginning with `-`, such as with a `--` separator.

ryanofsky commented:
Assuming we are ok with losing some backwards compatibility by starting to treating arguments that begin with `-` as options instead of non-options when they follow arguments that don't begin with `-`, there could be other interesting ways to extend this PR.

For example, `bitcoin-cli` could distinguish between options that begin with single and double dashes, and treat options after the RPC method name that begin with double dashes as RPC named parameters. In other words, accept:

```bash
bitcoin-cli createwallet mywallet --load_on_startup=1
```

as an equivalent to:


```bash
bitcoin-cli -named createwallet mywallet load_on_startup=1
```

This seems like it could provide a more convenient way of using named RPC parameters.

pablomartin4btc committed:
argsman, test: Double dashes as RPC named params

Treat options that begin wih double dashes as RPC named
params:

bitcoin-cli createwallet mywallet --load_on_startup=true

as an equivalent to:

bitcoin-cli -named createwallet mywallet load_on_startup=true

w0xlt reviewed (commented):
This command works on master but not in this PR:
`./build/bin/bitcoin-cli --datadir=/tmp/btc1 --signet getblockhash 1000`

Could it be related to GetCommandArgs()?

w0xlt reviewed (commented):
Restructuring `ProcessOptionKey` as below seems to solve the problem of `./build/bin/bitcoin-cli --datadir=/tmp/btc1 --signet getblockhash 1000`.

That seems to fix the `Method not found` regression while preserving the GNU-style options after the command and the new double-dash-named-params feature.

```cpp
bool ArgsManager::ProcessOptionKey(std::string& key, std::optional<std::string>& val, std::string& error, const bool found_after_non_option) {
    bool double_dash{false};
    const std::optional<std::string> original_val{val};
    std::string original_input{key};
    if (val) original_input += "=" + *val;

    // Normalize leading dashes
    if (key.length() > 1 && key[1] == '-') {
        key.erase(0, 1);   // `--foo` -> `-foo`
        double_dash = true;
    }
    key.erase(0, 1);       // `-foo`  -> `foo`

    KeyInfo keyinfo = InterpretKey(key);
    std::optional<unsigned int> flags = GetArgFlags('-' + keyinfo.name);
    const bool known_option = flags.has_value() && keyinfo.section.empty();

    // Handle the special "named RPC" case early ---
    if (double_dash && found_after_non_option && !known_option) {
        // Try to map this to `-named`, if supported by the binary.
        val.reset();
        KeyInfo named_keyinfo = InterpretKey("named");
        std::optional<unsigned int> named_flags = GetArgFlags('-' + named_keyinfo.name);

        if (named_flags && named_keyinfo.section.empty()) {
            // Binary supports -named: enable it and append the stripped parameter.
            std::optional<common::SettingsValue> named_value =
                InterpretValue(named_keyinfo, /* val */ nullptr, *named_flags, error);
            if (!named_value) return false;

            {
                LOCK(cs_args);
                m_settings.command_line_options[named_keyinfo.name].push_back(*named_value);
                // Strip the leading "--" before passing to the command as a named param
                m_command.emplace_back(original_input.substr(2));
            }
            return true;
        }

        // Binary doesn’t support -named: fall back and treat `--foo` as a
        // normal option (or unknown option handled below).
        keyinfo = InterpretKey(key);
        flags = GetArgFlags('-' + keyinfo.name);
    }

    // --- Normal option handling ---
    if (!flags || !keyinfo.section.empty()) {
        if (!found_after_non_option) {
            // Unknown global option: keep legacy behaviour (error)
            error = strprintf("Invalid parameter %s", original_input);
            return false;
        }

        // Unknown option after command: pass through to the command as an arg
        LOCK(cs_args);
        m_command.emplace_back(original_input);
        return true;
    }

    // Known option (including `--datadir`, `--signet`, etc.)
    std::optional<common::SettingsValue> value =
        InterpretValue(keyinfo, val ? &*val : nullptr, *flags, error);
    if (!value) return false;

    LOCK(cs_args);
    m_settings.command_line_options[keyinfo.name].push_back(*value);
    return true;
}
```

pablomartin4btc commented:
> This command works on master but not in this PR: `./build/bin/bitcoin-cli --datadir=/tmp/btc1 --signet getblockhash 1000` 
<detail>
<summarys>Thanks for testing it (not sure if that's valid, no test failed, perhaps we need to add one?), anyways just putting the output's diff here.</summarys><br>

`master`:
```
/build_master/bin/bitcoin-cli --datadir=/tmp/btc --signet getblockhash 1000
error code: -8
error message:
Block height out of range
```
this branch:
```
./build/bin/bitcoin-cli --datadir=/tmp/btc --signet getblockhash 1000
error: timeout on transient error: Could not connect to the server 127.0.0.1:8332

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.
Use "bitcoin-cli -help" for more info.
```
</details>

pablomartin4btc committed:
cli, test: Fix parsing of long (--) options

Improve handling of long (`--`) options passed to bitcoin-cli, such as
`--datadir` and `--signet`. Previously, some code paths in CLI didn't
properly recognize or forward double-dash options, causing unexpected
argument parsing behavior or timeouts.

This change ensures that long options are processed consistently,
matching ArgsManager behavior and CLI semantics.

Example fixed case:
    /build/bin/bitcoin-cli --datadir=/tmp/btc1 --signet getblockhash 1000

Co-authored-by: w0xlt <w0xlt@users.noreply.github.com>

pablomartin4btc committed:
argsman, refactor: Make ProcessOptionKey clearer

No behavior changed.

pablomartin4btc committed:
argsman, refactor: Make ParseParameters clearer

No behavior changed.

sedited commented:
@pablomartin4btc are you still working on this?

pablomartin4btc commented:
> @pablomartin4btc are you still working on this?

Yeah, working on a fix for the windows bug detected by @w0xlt. I'll update it soon.

maflcko commented:
Could turn into draft while CI is red?

pablomartin4btc committed:
argsman: Handle Windows slash options after commands

Allow command-specific options using '/' prefixes on Windows
after a command has been encountered.

Add a regression test covering '/D=windows' after a command.

Co-authored-by: w0xlt <w0xlt@users.noreply.github.com>

pablomartin4btc committed:
doc: Add release notes for GNU-style option parsing

pablomartin4btc commented:
-<ins>_**Updates**_</ins>:
- Rebased
   - Resolved conflicts with #34745 and incorporated the changes around guarding `cs_args` and the related helper refactors.
    - Resolved conflicts in `argsman_tests.cpp` with #28802 (92df785). Some tests relied on command-specific options appearing before the command; with the GNU-style parsing introduced in this PR, option ordering is no longer significant, so the affected tests were updated and additional cases were added.
- Added a fix for a Windows-specific bug where command-specific options using the `/` prefix were not handled correctly when specified after the command, together with regression test coverage. Thanks to @w0xlt for identifying the issue and proposing a fix (included as co-author).
- Added release notes.

w0xlt reviewed (commented):
`ArgsManager` now accepts options after non-option arguments, but `bitcoin-tx` still walked raw `argv` afterward. 

That means any option placed after the tx input could be parsed as an option and then incorrectly processed again as a transaction mutation command.

Suggested changes:

<details>
  <summary>diff</summary>

```diff
diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index f21db40f3c..7f94598d39 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -797,35 +797,42 @@ static int CommandLineRawTx(int argc, char* argv[])
     std::string strPrint;
     int nRet = 0;
     try {
-        // Skip switches; Permit common stdin convention "-"
-        while (argc > 1 && IsSwitchChar(argv[1][0]) &&
-               (argv[1][1] != 0)) {
-            argc--;
-            argv++;
+        std::vector<std::string> args;
+        if (const auto command{gArgs.GetCommand()}) {
+            args = command->args;
+        } else {
+            // Skip switches; Permit common stdin convention "-"
+            while (argc > 1 && IsSwitchChar(argv[1][0]) &&
+                   (argv[1][1] != 0)) {
+                argc--;
+                argv++;
+            }
+            args.assign(argv + 1, argv + argc);
         }
 
         CMutableTransaction tx;
-        int startArg;
+        size_t startArg;
 
         if (!fCreateBlank) {
             // require at least one param
-            if (argc < 2)
+            if (args.empty()) {
                 throw std::runtime_error("too few parameters");
+            }
 
             // param: hex-encoded bitcoin transaction
-            std::string strHexTx(argv[1]);
+            std::string strHexTx(args[0]);
             if (strHexTx == "-")                 // "-" implies standard input
                 strHexTx = readStdin();
 
             if (!DecodeHexTx(tx, strHexTx, true))
                 throw std::runtime_error("invalid transaction encoding");
 
-            startArg = 2;
-        } else
             startArg = 1;
+        } else
+            startArg = 0;
 
-        for (int i = startArg; i < argc; i++) {
-            std::string arg = argv[i];
+        for (size_t i = startArg; i < args.size(); i++) {
+            std::string arg = args[i];
             std::string key, value;
             size_t eqpos = arg.find('=');
             if (eqpos == std::string::npos)
diff --git a/test/functional/data/util/bitcoin-util-test.json b/test/functional/data/util/bitcoin-util-test.json
index 896decc4e3..03606149bf 100644
--- a/test/functional/data/util/bitcoin-util-test.json
+++ b/test/functional/data/util/bitcoin-util-test.json
@@ -56,6 +56,11 @@
     "output_cmp": "blanktxv2.json",
     "description": "Creates a blank transaction when nothing is piped into bitcoin-tx (output in json)"
   },
+  { "exec": "./bitcoin-tx",
+    "args": ["01000000000000000000", "-json"],
+    "output_cmp": "blanktxv1.json",
+    "description": "Parses an option after a transaction hex"
+  },
   { "exec": "./bitcoin-tx",
     "args": ["-create", "nversion=+1"],
     "return_code": 1,
```
</details>

w0xlt reviewed (commented):
The test can be improved by parsing `--datadir=...` with a fresh `ArgsManager` instead of using `ForceSetArg()`, so it exercises the real command-line path.

<details>
  <summary>diff</summary>
  
```diff
diff --git a/src/test/argsman_tests.cpp b/src/test/argsman_tests.cpp
index cab21ad4aa..1b5cf4d862 100644
--- a/src/test/argsman_tests.cpp
+++ b/src/test/argsman_tests.cpp
@@ -36,9 +36,16 @@ BOOST_AUTO_TEST_CASE(util_datadir)
     args.ClearPathCache();
     BOOST_CHECK_EQUAL(dd_norm, args.GetDataDirBase());
 
-    args.ForceSetArg("--datadir", fs::PathToString(dd_norm) + "/");
-    args.ClearPathCache();
-    BOOST_CHECK_EQUAL(dd_norm, args.GetDataDirBase());
+    {
+        ArgsManager parsed_args;
+        parsed_args.AddArg("-datadir=<dir>", "", ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION, OptionsCategory::OPTIONS);
+        const std::string datadir_arg{"--datadir=" + fs::PathToString(dd_norm) + "/"};
+        const char* argv_datadir[] = {"ignored", datadir_arg.c_str()};
+        std::string error;
+        BOOST_CHECK(parsed_args.ParseParameters(2, argv_datadir, error));
+        BOOST_CHECK_EQUAL(error, "");
+        BOOST_CHECK_EQUAL(dd_norm, parsed_args.GetDataDirBase());
+    }
 
     args.ForceSetArg("-datadir", fs::PathToString(dd_norm) + "/.");
     args.ClearPathCache();
```

</details>

pablomartin4btc commented:
> ... `bitcoin-tx` still walked raw `argv` afterward.
> 
> Suggested changes...

Thanks, it makes sense. I agree `bitcoin-tx` should consume the already parsed command args instead of walking raw `argv` again, otherwise options accepted after non-options can be processed twice. I’ll update it and add the regression test as well.

pablomartin4btc commented:
> The test can be improved by parsing `--datadir=...` with a fresh `ArgsManager` instead of using `ForceSetArg()`, so it exercises the real command-line path.

I agree, better to have it parsed, I'll update the test as well. Thanks!

pablomartin4btc committed:
bitcoin-tx: Use parsed command arguments from ArgsManager

Use the command arguments already parsed by ArgsManager instead
of re-processing raw argv. This ensures options specified after
non-option arguments are handled consistently.

Co-authored-by: w0xlt <w0xlt@users.noreply.github.com>

pablomartin4btc commented:
-<ins>_**Updates**_</ins>:
- Addressed @w0xlt feedback and incorporated the suggested changes:
  - [(1)](https://github.com/bitcoin/bitcoin/pull/33540#pullrequestreview-4431441001) Updated `bitcoin-tx` to consume the command arguments already parsed by `ArgsManager`, avoiding re-processing raw `argv`.
    - Added regression test coverage (`bitcoin-util-test.json`) which fails if the `bitcoin-tx` changes are reverted (eg `./build/test/functional/tool_utils.py --loglevel=DEBUG`).
  - [(2)](https://github.com/bitcoin/bitcoin/pull/33540#pullrequestreview-4431905598) Updated the `--datadir=...` test in `argsman_tests.cpp` to use `ArgsManager::ParseParameters()` instead of `ForceSetArg()`.

pablomartin4btc commented:
-<ins>_**Updates**_</ins>:
  - Updated `bitcoin-cli` help text to clarify that options can be specified before or after commands (no behavioural change).

ryanofsky commented on src/common/args.cpp:
```diff
@@ -191,15 +192,22 @@ bool ArgsManager::ProcessOptionKey(std::string& key, std::optional<std::string>&
+        if (!found_after_non_option) {
+            error = strprintf("Invalid parameter %s", original_input);
+        } else {
+            // if the option is invalid but comes after a non-option (found_after_non_option)
+            // leave it up to the command if it accepts args that begin with "-"
```
In commit "argsman, cli, test: Allow options after non-option arguments" (37f03fac5fc46716e339eab53e2a463303a8e2c5)

I think this behavior is problematic, and would suggest dropping this `else` block and always treating invalid options as errors, regardless of whether they appear before or after commands.

The problem with this is that it's fragile: it makes it too easy to misspell an option after a command (e.g. "-walllet=foo") and have it parsed as a positional argument that could trigger unintended behavior, when the same misspelling in a different position would result in a clear error. It would be safer to just consistently raise errors when it looks like a bad option has been passed.

This behavior is also a problem because it's not forward-compatible: if new options are added, command arguments that were previously passed as positional arguments could now be parsed as options.

I understand reason for taking this approach is to provide backwards compatibility, so `createwallet -mywallet` still creates a wallet called "-mywallet" after this commit just like it does currently. But providing this backwards compatibility doesn't seem worth the decreased safety, increased complexity, and loss of forward compatibility it brings.

Also worth noting that current commit doesn't provide backwards compatibility in cases where valid option names are used. For example `createwallet -regtest` currently creates a wallet called "-regtest" and after this commit it would no longer do that.

I think the best approach is to treat options consistently and always treat unrecognized options as errors.

To support passing positional arguments that begin with dashes, a good approach is to follow the getopt convention, accepting a literal `--` separator that treats all arguments after the separator as positional arguments. So `createwallet -- -mywallet` and `createwallet -- -regtest` would both create the specified wallets with leading dashes.

ryanofsky commented on src/common/args.cpp:
```diff
@@ -210,6 +238,10 @@ bool ArgsManager::ProcessOptionKey(std::string& key, std::optional<std::string>&
 
+    if (double_dash && keyinfo.name =="named") {
+        m_command.emplace_back(original_input.substr(2));
+    }
```
In commit "argsman, test: Double dashes as RPC named params" (ba1e8863077c36cedd92ba270cb0bf6090fb4d90)

This approach which enables the `bitcoin-cli` `-named` option and strips leading double dashes off of `bitcoin-cli` arguments is clever, but I don't think it is a maintainable solution. There shouldn't really be `bitcoin-cli` specific code inside `ArgsManager`. And the behavior of `-named` is not really what we want because double dash arguments should always to be passed as RPC named arguments and `-named` will sometimes pass them as positional arguments instead (depending on whether they contain `=`, and the parameter position, and whether the parameter name is recognized).

I think a better solution would be to have `bitcoin-cli` configure `ArgsManager` to treat double-dash options that appear after commands as passthrough options that can be retrieved in a `vector<pair<string, optional<string>>>` and added directly to the JSON request.

It may be better to save this change for a followup PR to reduce complexity here.

ryanofsky reviewed (commented):
Code review e47c722324dd33337eff4c0a24bb051daebe5f27. These are nice changes! The ability to specify options after commands seems useful and more natural than the current way of passing subcommand options. Having the ability to pass named RPC parameters more conveniently without `-named` also seems great.

My main feedback is that I think specifying invalid options after commands should still be errors. If options are now going to be accepted after commands, it makes sense to validate them the same way as they are validated when they appear before commands, even if this means there is a little less backwards compatibility in some edge cases, and some things that used to work will now be "Invalid parameter" errors. I wrote more detailed comment about this below.

Also it looks like internally some cleanups could be made. It seems like some commits (5af80d0447dabd1a773de5e4f11a9fb88cf14bde, be8573a9ef4fc500b6bcfd7c797fa75f052568ec) may be fixing bugs introduced in earlier commits? I think in order for this to be merged, internal commits need to be bug-free. 

It also looks like some commits refactor code that previous commits introduced, when it would seem better to refactor first, so the same code isn't changing multiple times within the PR.

Motivation for some of the refactorings also seems a little hazy, and there are some strange things like input arguments being passed by mutable reference when const references would be more appropriate.

Not sure if it's too much to ask, but I'd very interested in seeing a much smaller version of this PR that that doesn't do any refactoring at all and doesn't implement the RPC named argument extension. If there could be a smaller spinoff PR that **only** adds the ability to accept options specified after commands, and supports a `--` separator to allow non-option arguments to begin with dashes, I feel like that would be a much easier starting point for review, and the other changes could be followups.

pablomartin4btc commented:
Putting this PR on draft in favor of #35831— a simpler, focused PR that implements only the core change @ryanofsky requested above:
- options recognized after non-option arguments (GNU-style),
- `--` separator support,
- and unrecognized options always being errors.

The refactoring, RPC named params via double-dash, and bitcoin-tx changes are all dropped.
